### ISSUE TITLE
(aws-apigateway): RestApi imported with method fromRestApiAttributes not updating current deployment in stage

### ISSUE THREAD

**Issue by evilsephiroth** [aws/aws-cdk#13526]
I've imported from a serverless project this restApi.
```
const restApiId = cdk.Fn.importValue(envRelated(`ApiGateway-restApiId`));
const rootResourceId = cdk.Fn.importValue(envRelated(`ApiGateway-rootResourceId`));
this.apiGw = apigateway.RestApi.fromRestApiAttributes(this, envRelated(`apigw`), {
	restApiId: restApiId,
	rootResourceId: rootResourceId,
});
```
When deploying with cdk deploy, all resources are updated but not the stage.
No deployment is created.
I've tried also to force a deployment with this snippet.
```
new Deployment(this,envRelated('deployment-stage') , {
      api: apiGw
    });
```
Salting the deployment id makes no difference.
After the deployment is done, the api gateway is still pointing to the previous deployment instead of the new deployment done.

![chrome_sGUK4TBoqT](https://user-images.githubusercontent.com/199461/110676506-52f88280-81d4-11eb-89b7-e12889627b5a.png)

Specifying the stage with the forced deployment fails because stage already exists.

### Environment

  - **CDK CLI Version:** 1.9.2
  - **Module Version:** 1.9.2
  - **Node.js Version:** 12.14.1
  - **OS:** Ubuntu Linux 20.04
  - **Language: ts

**Comment by baumannalexj** [comment]
```
const apiDeployment = new cdkApiGateway.Deployment(this, `api-deployment`, {
            api: <importedApi>,
            description: `maybe put datetime or artifact version here`,
            retainDeployments: true,
        });
        apiDeployment.addToLogicalId(cdkConfig.DATETIME_ISO); //need force change deployment hash to force new deployment

```

and for a specific stage
```
        apiDeployment.resource.stageName = cdkConfig.ENV_UPPER;          // https://stackoverflow.com/a/64369331/6791342 // if the 'stageName' already exists (from the core apigateway deployment) then the existing stage will be used !


```

**Comment by evilsephiroth** [comment]
> @evilsephiroth try this
> 
> you need to bust the deployment sha with `addToLogicalId`
> 
> ```
> const apiDeployment = new cdkApiGateway.Deployment(this, `api-deployment`, {
>             api: <importedApi>,
>             description: `maybe put datetime or artifact version here`,
>             retainDeployments: true,
>         });
>         apiDeployment.addToLogicalId( new Date().toISOString() ); //need force change deployment hash to force new deployment
> ```
> 
> and for a specific stage
> 
> ```
>         apiDeployment.resource.stageName = <your stage name>;          // https://stackoverflow.com/a/64369331/6791342 // if the 'stageName' already exists (from the core apigateway deployment) then the existing stage will be used !
> ```

Added the logicalId but no change. Still the deployment occurs but it's not selecting the latest deployment in the stage.
And apideployment.resource.stageName won't compile in 1.92 version.

**Comment by NGL321** [comment]
Hey @evilsephiroth,

I am unsure as to why this was not correctly marked as a bug report, but I have now corrected it. Someone from the team will attempt to reproduce your issue as soon as able. Please let us know if you have any updates, and we will update this issue when there is an opportunity to investigate further.

**Comment by nemanjko** [comment]
@evilsephiroth I have the same issue. 

Unable to run new deployment to update a stage in API Gateway. Tried previous suggestion by @ baumannalexj but still the same. 

The workaround I use is to run Lambda custom resource and trigger the deployment explicitly: 

    response = boto3.client.create_deployment(
        restApiId='hvgcqh4dgk',
        stageName='device4'
    )

**Comment by anun-demo** [comment]
> Hey @evilsephiroth,
> 
> I am unsure as to why this was not correctly marked as a bug report, but I have now corrected it. Someone from the team will attempt to reproduce your issue as soon as able. Please let us know if you have any updates, and we will update this issue when there is an opportunity to investigate further.

We're having the exact issue as well

**Comment by gawdzer** [comment]
> Hey @evilsephiroth,
> 
> I am unsure as to why this was not correctly marked as a bug report, but I have now corrected it. Someone from the team will attempt to reproduce your issue as soon as able. Please let us know if you have any updates, and we will update this issue when there is an opportunity to investigate further.

We're having the same issue as well :[

**Comment by nija-at** [comment]
Sounds like the same as https://github.com/aws/aws-cdk/issues/12417. 

You will need to manually salt your deployment every time any resource changes. This is a limitation of the `fromRestApiAttributes()` method.

**Comment by nija-at** [comment]
Here's how to salt your own deployment - 

```ts
const logicalId = // salted id
const cfnDeployment = deployment.node.defaultChild as CfnDeployment;
cfnDeployment.overrideLogicalId(logicalId);
```

The value of the logicalId here should be such that it changes every time a Resource or Method is modified, and this stack must be deployed last.

**Comment by gawdzer** [comment]
> Here's how to salt your own deployment -
> 
> ```ts
> const logicalId = // salted id
> const cfnDeployment = deployment.node.defaultChild as CfnDeployment;
> cfnDeployment.overrideLogicalId(logicalId);
> ```
> 
> The value of the logicalId here should be such that it changes every time a Resource or Method is modified, and this stack must be deployed last.

Changing the value of the logicalId does not affect the result: One can create a new deployment, but it's impossible to set is as the 'active' deployment for an already-existing stage.

**Comment by hassanazharkhan** [comment]
@gawdzer @nemanjko Since you both were facing the same issue, Did you have any luck with the fix of this one?

I'm facing the same issue as well now, I even tried every possible solution for this but no luck so far, any help would be greatly appreciated

**Comment by hassanazharkhan** [comment]
@otaviomacedo So just to sum up there is no way to set the Deployments to the current stage if the RestApi is imported using CDK?

FYI, I'm able to set the current deployment using the console

**Comment by montoyan877** [comment]
Any news on this topic?

**Comment by Simonl9l** [comment]
@otaviomacedo, @hassanazharkhan, @nija-at, @NGL321  et  al - I have a similar related unresolved issue - per [this](https://stackoverflow.com/questions/63950199/how-to-use-an-existing-stage-in-api-gateway-deployments-in-aws-cdk/64369331#64369331) it seem a way to deploy to an existing stage is via `deployment.resource.stageName = 'YourStageName';`

However in dotnet/C# this property is not accessible.

 I also note that Stage.FromStageAttributes per the CDK dotnet [documentation](https://docs.aws.amazon.com/cdk/api/v2/dotnet/api/Amazon.CDK.AWS.APIGateway.Stage.html#Amazon_CDK_AWS_APIGateway_Stage_FromStageAttributes_Constructs_Construct_System_String_Amazon_CDK_AWS_APIGateway_IStageAttributes_) does not seem to exist, at least in version 2.43.1, not that this would help but a divergence of assembly vs documentation.

I'm essentially importing the RestApi into a number of different stacks via `RestApi.FromRestApiAttributes`, adding Resources as applicable for that stack and then need to deploy it, in each case; but see no way to do it specifically in DotNet/C#

I need to redeploy the same stage each time but get the error `prod already exists in stack`.

**Comment by minhhuynh-smg** [comment]
Why this topic was closed? This issue seems to be still a valid request for the feature of imported RestApi.

**Comment by Simonl9l** [comment]
@minhhuynh-smg I’d tagged team members per the bot instructions to no avail. Someone probably need to “open new issue referencing this one”

**Comment by muzeke** [comment]
Wondering if anyone has ever got a working solution for this problem as We are currently experiencing the same issue. One thing i noticed though, looking at the deployment history it seems that it always points to the latest deployment but my changes on the api resources and methods are not working when i test the endpoints. (I will either have to manually do deploy in aws console UI or do cdk deploy twice to make it work) 

```typescript
const deployment = new apigateway.Deployment(
    scope,
    `Deployment-${new Date().toISOString()}`,
    {
      api: restApi,
    }
  );

  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  deployment.resource.stageName = 'main';

  const logicalId = Date.now().toString(); // salted id
  deployment.addToLogicalId(logicalId);
```

**Comment by jmihalich** [comment]
Seriously guys, I'm tired of this.  CDK Development is 95% google search, and 5% actual coding.  The documentation is horrendous, you really need to make it a priority to update it.  The above almost worked for me.  The issue, if you look at the template is that the deployment is created in parallel with everything else if there is no dependency on anything, and winds up running first before the actual new resource or whatever is applied to an existing gateway.  I couldn't depend on the "FromResourceAttribute" object cuz it didn't add a dependency.  I had to have the Deployment depend on the new resource I added. For those who find this, here is example .Net code that worked for me.  You can convert this to the language of your choice:

`
        var stack = new Stack(scope, id, stackProps);

        var function = new Function(stack, "TestFunction", new FunctionProps()
        {
		    //NOTE: Substitute withyour own new lambda code
            FunctionName = "MyTestFunction",
            Runtime = Runtime.PYTHON_3_10,
            Code = Code.FromAsset("lambda"),
            Handler = "lambda_function.lambda_handler"
        });

        var restApi = RestApi.FromRestApiAttributes(stack, "ExistingRestApi2", new RestApiAttributes()
        {
		    //You have to export the api-gateway-api-id and api-gateway-root-resource-id values as CfnOutputs from your main stack that initially deployed the gateway.
            RestApiId = Fn.ImportValue("api-gateway-api-id"),
            RootResourceId = Fn.ImportValue("api-gateway-root-resource-id")
        });

        var stage = Amazon.CDK.AWS.APIGateway.Stage.FromStageAttributes(stack, "PublicAPIStage", new StageAttributes()
        {
            RestApi = restApi,
			
			//Replace with your existing stage name
            StageName = "public-api"
        });

        var newResource = restApi.Root.AddResource("graph");
        newResource.AddProxy(new ProxyResourceOptions()
        {
            DefaultIntegration = new LambdaIntegration(function),
            AnyMethod = true,
            DefaultMethodOptions = new MethodOptions()
            {
                AuthorizationType = AuthorizationType.NONE
            }
        });
        newResource.Node.AddDependency(function);

        var deployment = new Deployment(stack, "ApiTestDeployment", new DeploymentProps()
        {
            Api = restApi,
            RetainDeployments = false,
        });
		
		//this salts the deployment
        deployment.AddToLogicalId(DateTime.UtcNow.ToString());
		
        //make the deployment dependent on the resource so it runs last
        deployment.Node.AddDependency(newResource);

        //StageName is unavailable in the L2 deployment class, so you need to get the L1 resource.
        var cfnDeployment = (CfnDeployment)deployment.Node.DefaultChild!;
		
		//set the name of the stage on the L1 deployment object.
        cfnDeployment.StageName = "public-api";`

**Comment by defaultusername8** [comment]
For anyone who might have the same problem:

after some experimenting, just modifying the ` latestDeployment` logical id worked for me without creating my own deployment. I suspect that might not always work (in my case, I was doing a change to only lambda integration properties and it was not ending up in a apigw deployment at all - not only was the deployment not active, but it was not even created), but it's worth a shot before messing with creating your own deployment and having the `deploy: true` option bug out.

code:
{
// messing around with api properties that I assume do not end up in the hash
}
const apiGateway = new SpecRestApi(
    scope,
    `MyRestApiGateway`,
    {
      apiDefinition: ApiDefinition.fromInline(apiDefinitionJSON),
      restApiName: `MyRestApiGatewayName`,
      deploy: true
      ...
     }
  apiGateway.latestDeployment?.addToLogicalId(Date.now());
